### DISCUSSION TITLE
Theme override on a per album basis

### DISCUSSION THREAD

**Discussion by TT-ReBORN** [TT-ReBORN/Georgia-ReBORN#112] [Ideas]
[acstrp](https://github.com/acstrp)
Any thought on having a theme override on a per album basis?

I stick with Reborn for the most part, but some album covers really pop with one of the neon themes or with one of the Reborn Fusion options.

Would definitely be cool to be able to set an override for specific albums with all the currently selected Theme and Style settings - right click on cover art and have an option to "set preferred theme/style to current selection" and a "clear preferred theme/style" option to remove it.

Anyway, just an idea since I find myself manually adjusting theme and style for some albums just for the cool visuals.

---

[TT-ReBORN](https://github.com/TT-ReBORN)
I don't know how you would even implement this and how it would actually work.
What you can actually do, is to actually set a custom preset.
For this, open your georgia-reborn-custom-config.jsonc file and look for the section "customStylePreset": 
After you've all set it, go to Options > Style > Presets > Use preset > User settings
-TT

---

[acstrp](https://github.com/acstrp)
I'm not sure a custom preset would work on a per album basis?

**Specific example:**

Say you have album A, B, C, D, E and your theme is Reborn. But album B has a cover art that looks amazing with the Neon Red theme, and album D looks incredible with Reborn Fusion + bevel + gradient.

You can manually switch themes when album B and D play, then back to the Reborn theme each time.

But I am imagining an auto switch to Neon Red when album B plays, then back to the selected theme of Reborn after it ends, and the same switch to Reborn Fusion when album D plays.

**From a user perspective:**

When album B plays, the user can switch to Neon Red once, right click the cover art, select "set album preferred theme/style to current selection" then switch back to the Reborn theme. But now any time a song from that album plays, the theme will auto switch to Neon Red, and when a different album plays, the theme will auto switch back to the Reborn theme (or if that album has an override set, the specified override theme). A user can clear an album override with a right click on the cover art and select "clear album preferred theme/style."

So the 5 albums above played sequentially would be album A - Reborn theme active, album B - Neon Red theme active, album C - Reborn theme active, album D - Reborn Fusion + bevel + gradient active, album E - Reborn theme active -- all with no user intervention after selecting the 2 overrides.

**Implementation wise:**

The simplest way I can think of without involving a database is to create an override folder, and when an override is set, create a config text file named for that artist+album containing either the preset name or the actual theme values.

At every song play, if an override file exists for that artist+album, load the theme as specified in the override, else continue with the active theme.

If an override is cleared, delete the associated config text file.

Not sure if there is a better way to do it, but this way doesn't involve a database or enumerating through a potentially long override list.

**In conclusion:**

Is this something anyone else would even like? I have no idea lol. It was just a thought. Thanks for reading this overly long block of text.

**Comment by TT-ReBORN** [comment]
@acstrp, I've moved this topic from the beta thread here and we will continue the discussion here. 
It does sound like a interesting concept and I will consider it in future updates ( after the first public release ).
I still have no idea or blueprint, how it could be implemented to make it work properly. 
I would need to think about it when the right time comes, this is complicated stuff.
When does an album art look good ( subjective ) and what theme and style combo would look best for this particular cover, etc.

-TT

**Comment by kbuffington** [comment]
Georgia doesn't have themes per-se, but since the themes are generated from a specific color it does allow the individual file to decide what color works best. i.e.: a tag of "THEMECOLOR" with a value of `"rgb(84, 145, 182)"` will then use that color and not any of the colors it found in the image.

I still have plans at some point to move that from tagging in the individual files to a metadb record which could be set per album which would allow users to easily specify a color and not have to alter the actual file data.

You could do something similar for a specific theme, but then you'd have to have a non-changing way to reference them. Either an immutable index, or some kind of naming system that allows for easy lookup and fails over properly if the theme can't be found. Just a thought.

**Reply by TT-ReBORN** [reply]
The `%THEMECOLOR%` never caught my attention in your code, but NOW I know what it's supposed to do and 
what you meant by color hardcoded :sweat_smile: . That's some creative way of abusing tags in music files! 
So if I encounter some music files and see this tag, I know it was you :love_letter:

**Comment by acstrp** [comment]
@TT-ReBORN Thanks for moving my comment to a new thread, totally should have started it here anyway.

I only got the idea since you created so many possible theme/style options and combinations. I had been cycling through lots of options for different albums and making notes of which settings I liked best for each album, so of course I thought.... well, it would be great if I didn't have to manually switch if I really liked one theme/style best for a particular album.

I wasn't so much thinking of specific granular color overrides, but just a preference for one of the existing settings or presets. The subjective part of what looks good is such a personal thing, so that was why I thought a user being able to force a theme/style for an album would be a nice option.

I didn't really consider how the code works when I thought of this, but I was thinking whatever mechanism switches from say... Reborn + Bevel + Blend to Neon Red when user manually makes the switch could automatically be called when a song from a specific album plays; unsure if that would be feasible.

Anyway, thanks for considering this at some point in the future. It's really just a "this would be cool to have!" type of thing.

**Comment by TT-ReBORN** [comment]
So yesterday when I actually thought about this more in depth, I had this epiphany...

I don't know if you already know about this, but by default if you double click somewhere on the lower bar, it will choose a random of 
88 different style preset I've made. These style presets contain a chosen theme with its different style presets. The action double click on lower bar is a default setting and can be changed via Options > Style > Presets > Auto random > New album ( for this case ).
You can also filter out themes you don't want to be in the random picker pool via Options > Style > Presets > Auto random > Use presets.


Now we don't really need a complicated config setup or database for this to work, 
but I can actually make a new "harmonic controlled random preset picker" that is based several color rules, for example:

- When the album art is very dark ( mostly dark or black colors ), choose a random Black preset or Neon theme preset.
- When the album art is very bright ( mostly light or white colors ), choose a random White theme preset.
- When the album art has mostly colors from the middle spectrum, choose a random Reborn theme preset.
- When the album art has mostly colors from the middle spectrum and primary and secondary chosen color has a wide 
  color distance ( big distinction between those two ), choose one of the three Reborn fusion presets.

This would be the easiest way to achieve something similar what you have requested.

-TT

**Reply by acstrp** [reply]
Already been enjoying the "lower bar double click" to get a random theme for a while now, good stuff!

A harmonic color picker for a random theme is definitely an interesting idea. That would still require a manual selection though, which I was hoping to avoid. Double clicking in the lower bar would definitely be quicker than going through the menus.

I have some album covers with lots of colors but a black border that really look good with a neon theme, but may still not have enough dark color overall for a harmonic selection to come out with the theme I would pick.

I'll keep pondering!

**Comment by TT-ReBORN** [comment]
I bring good news @acstrp, I have implemented this now =)

Actually the solution was in front of me the whole time, @kbuffington's great idea to use custom embedded tags.

You now have the ability to use custom theme, styles and preset for every album, so how about you test it now? 
You need to [download from the master](https://github.com/TT-ReBORN/Georgia-ReBORN/archive/refs/heads/master.zip) and see in the FAQ under Customization section:
["How can I set custom theme, styles, preset for music albums ( will be available in the upcoming version ) ?"](https://github.com/TT-ReBORN/Georgia-ReBORN/discussions/58)
Everything you need to know is described there.

-TT

**Reply by acstrp** [reply]
You did it!!

Thank you for implementing my (crazy) suggestion!

I haven't tested it out thoroughly yet but everything works as intended with the couple albums I added tags to. I intend on just using presets for the albums I'd like a specific theme for, but I will try and test out all the options for good measure.

Here is the first album that inspired my suggestion.

It is not bad at all with the Reborn theme:
![reborn](https://github.com/TT-ReBORN/Georgia-ReBORN/assets/36980498/a73ff426-9e99-4739-a0c0-6cbe0cff8298)


Harmonic gave me a couple options that went in the opposite direction, definitely prefer these less than plain Reborn:
![harmonic1](https://github.com/TT-ReBORN/Georgia-ReBORN/assets/36980498/36392afc-d42e-49e8-a6f8-6bc3f6945c36)
![harmonic2](https://github.com/TT-ReBORN/Georgia-ReBORN/assets/36980498/806293a6-175e-44bf-9c17-3ec756737ff3)


But now with the override I am going with a simple Neon Red+Bevel which suits the album just right to me (and meshes perfectly with the text on the cover):
![override](https://github.com/TT-ReBORN/Georgia-ReBORN/assets/36980498/2db02b3b-6b23-42e6-a328-3efcbe734aaf)

Ahh, just right indeed.

**Reply by TT-ReBORN** [reply]
Glad you like it =) 

It was a great idea to be honest, so thanks for that. Now if you're really crazy, you could tag every album 
with custom themes and styles or use a theme preset, i.e maximum freedom and customization.

I still think most users will be interested with using the harmonic mode with a theme preset override, it makes a great combo.

**Reply by acstrp** [reply]
Not that crazy, hah! Reborn and Harmonic mode do pretty great most of the time so it is really not necessary.

Pretty awesome you ended up designing Harmonic mode out of this as well.

Next albums getting an override: Guns N' Roses - Appetite For Destruction and Def Leppard - Hysteria.

**Comment by TT-ReBORN** [comment]
I have implemented two new cool features that makes your life much easier when tagging your albums with the GR-tags.
- Auto-write GR-tags to your music files.
- Theme sandbox mode

There is actually no more need to look up for values and manually set all GR-tags, you can also restore your theme state.

Taken from the updated FAQ:

> You can set your preferred theme, styles or a preset and select the album in the playlist or library.
> Right click for context menu and select the option "Write theme tags". Your album will now have all theme tags written into your files.
> 
> Before you write your albums with theme tags, you should try the theme sandbox mode ( Options > Settings > Theme sandbox ).
> It is useful when finishing up with tagging your albums and you want to restore your original theme state. When this mode is active,
> you can change different themes, styles or presets and after disabling the theme sandbox mode, previously used theme settings
> can be restored.

You can try it out now, it's been implemented in the [master](https://github.com/TT-ReBORN/Georgia-ReBORN/archive/refs/heads/master.zip).

-Tom

**Reply by acstrp** [reply]
Loving the theme overrides, and the menu option to write theme tags without looking up specific values.

This isn't  a bug at all, just something I noticed -

When double clicking near track title for a random theme, if you get one you really like that happens to be a random preset... you can't automatically write the tags for that exact specific theme combination, you get the random theme tags written to file, which is correct since it is the current theme - but the next time it loads you will likely get a similar but different look. It might be nice to be able to tag that specific color combination when using the random selector.

Anyway, it was just something that jumped out at me.

**Reply by TT-ReBORN** [reply]
Nice to hear that you like the new features, 

---

Sorry but I do not really understand what you mean. When double clicking on the lower bar, it will select 
a random preset, harmonic preset or theme preset based on active `Options > Preset > Select mode`.
You can then use "Write theme tags" and it will write the %GR_PRESET% to your music file. 
When any GR-tags exist in your music files, you can not use the lower bar double click to generate 
a new preset because they have a higher load priority, the double click generate a new preset action will be ignored.

Btw, when using a theme preset, only the %GR_PRESET% tag will be written. 
%GR_THEME% and %GR_STYLE% will not be written ( they are not necessary ) because %GR_PRESET% already has both.

**Reply by acstrp** [reply]
I just meant a random "random preset" will give you multiple possible color combinations.

If you double click at the bottom for a random theme and end up with say... Preset Random Beveled Auto Dark and tag it, you won't always get the same colors the next time - ie. you are tagging the random preset vs the specific colors at that time. Maybe just point that out in the documentation?

After thinking about it, really doesn't seem worthwhile to be able to tag the specific colors a random preset generated.

**Reply by TT-ReBORN** [reply]
Aha now I understand, you mean when the Random theme ( Options > Theme > Random ) is active.
The Random theme always generates a new random color when playing a new album. When using the lower bar double click
and a Random theme preset was selected, it does not save its theme colors, because it is not implemented. 

That means the unused %GR_THEMECOLOR% tag will finally be useful when the Random theme is active. 
I can implement a function to write the current displayed Random theme color to the tag, then you will always have
the color with its styles.

---

P.S Many people are underestimating the Random theme in general, but when used correctly it's amazing.
Here is a tip: Select the Random theme via `Options > Theme > Random` and deactivate all Select presets via
`Options > Preset > Select presets > Deactivate all `, go again and activate `Options > Preset > Select presets > Random`.

Now this will randomly only select the Random presets when double clicking on the lower bar. 
With every double click, it will also generate a new random theme color with a Random theme preset.
You can play with this method on different albums ( they have different color brightness, i.e different blending )
and achieve some great results.

**Reply by acstrp** [reply]
Totally agree on the Random themes, really get some excellent color combinations I would not have thought of otherwise.

I was backpedaling a bit in my last post - I realized I was essentially making another feature request, but I know how much you are working on lately so I am trying not to ask for anything new.

Looking forward to it if/when you get around to it!

**Reply by TT-ReBORN** [reply]
OK, I've implemented this. It will save the primary color as %GR_THEMECOLOR% tag to your music files but ONLY 
when using the Random theme.

Just activate Random theme and use the new feature ( right click on album art for context menu > Generate new color ) 
that I have also added to generate a color you like and then use "Write theme tags".

[This is available now in the master.](https://github.com/TT-ReBORN/Georgia-ReBORN/archive/refs/heads/master.zip)

**Reply by acstrp** [reply]
"Generate new color" is a great idea!

Not sure it is 100% working as intended - I was double clicking in the bottom bar to cycle through some options, then I wrote the tags to file.

The tags written were:

%GR_PRESET%   randomP04
%GR_THEMECOLOR%   rgb(52, 30, 51)

But like you said, because the theme isnt reborn, new colors are generated for each track on playback instead of using the tagged colors.

**Reply by TT-ReBORN** [reply]
I have simplified the behavior of the Random theme when generating new colors and it should also work when using theme presets.

When you're using the Random theme and "Write theme tags" you can still use the new "Generate new color" feature, but you can not use the lower bar double click to generate a new theme preset. This is because the music file has the embedded GR-tags and they have the highest priority, otherwise it would not work. So if you want to generate a new random theme preset, you would first need to delete all embedded GR-tags in your music files of THAT album.

[This is available now in the master.](https://github.com/TT-ReBORN/Georgia-ReBORN/archive/refs/heads/master.zip)
